acorn/src/expression.js

```diff
@@ -178,16 +178,27 @@ pp.parseExprOps = function(noIn, refDestructuringErrors) {
 // defer further parser to one of its callers when it encounters an
 // operator that has a lower precedence than the set it is parsing.
 
-pp.parseExprOp = function(left, leftStartPos, leftStartLoc, minPrec, noIn) {
+// shortCircuitOps:
+// * 1 = only `||` and `&&` are allowed
+// * 2 = only `??` is allowed
+// * others = all
+
+pp.parseExprOp = function(left, leftStartPos, leftStartLoc, minPrec, noIn, shortCircuitOps) {
   let prec = this.type.binop
   if (prec != null && (!noIn || this.type !== tt._in)) {
+    let logical = this.type === tt.logicalOR || this.type === tt.logicalAND
+    let coalesce = this.type === tt.coalesce
+    if (shortCircuitOps === 1 && coalesce || shortCircuitOps === 2 && logical) {
```

Comment:
Hmm these magic numbers are not very clear tbh.

Reply:
Thank you for your review. I have updated it.